### PR TITLE
[SPARK-34970][SQL][SERCURITY] Redact map-type options in the output of explain()

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

The `explain()` method prints the arguments of tree nodes in logical/physical plans. The arguments could contain a map-type option that contains sensitive data.
We should map-type options in the output of `explain()`. Otherwise, we will see sensitive data in explain output or Spark UI.
image

### Why are the changes needed?

Data security.

### Does this PR introduce _any_ user-facing change?

Yes, redact the map-type options in the output of `explain()`

### How was this patch tested?

Unit tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 1]
```diff
@@ -69,6 +71,18 @@ trait ExplainSuiteHelper extends QueryTest with SharedSparkSession {
   protected def checkKeywordsExistsInExplain(df: DataFrame, keywords: String*): Unit = {
     checkKeywordsExistsInExplain(df, ExtendedMode, keywords: _*)
   }
+
+    /**
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:75]: indentation?

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 2]
```diff
@@ -346,6 +360,45 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
         Nil: _*)
   }
 
+  test("Redact Map type options in explain output") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:363]: Shall we add `SPARK-34970: ` prefix?

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 3]
```diff
@@ -346,6 +360,45 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
...
+    Seq(SimpleMode, ExtendedMode, FormattedMode).foreach { mode =>
+      checkKeywordsExistsInExplain(cmd, mode, value)
+    }
+    Seq(SimpleMode, ExtendedMode, CodegenMode, CostMode, FormattedMode).foreach { mode =>
+      checkKeywordsNotExistsInExplain(cmd, mode, password)
+      checkKeywordsNotExistsInExplain(cmd, mode, token)
+    }
+  }
+
+  test("Redact CaseInsensitiveMap type options in explain output") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:380]: ditto.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with a few minor comments)

**Review by maropu — APPROVED** [review]
Looks fine except for the @dongjoon-hyun comments.

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by mridulm** [comment]
+CC @xkrogen

**Comment by gengliangwang** [comment]
GA passes. Merging to master and branch-3.1.

**Pull request closed without merging**